### PR TITLE
JBEHAVE-1537 Add ability to load aliases from resources

### PR DESCRIPTION
**Pull request by uarlouski**

### CODE REVIEW

File: distribution/src/site/content/aliases.html [hunk 1]
```diff
@@ -73,6 +73,29 @@ <h2>Aliases</h2>
...
+    [
+        {
+            "name": "When the item price is $price",
+            "aliases":
+            [
+                {
+                    "name": "When the item price becomes $price"
+                },
+                {
+                    "name": "the item price equals to $price"
```
Comment (valfirst) [inline distribution/src/site/content/aliases.html]:
```suggestion
                    "name": "When the item price equals to $price"
```

File: examples/core/src/main/java/org/jbehave/examples/core/stories/alias.story [hunk 2]
```diff
@@ -0,0 +1,10 @@
+Scenario: Verify aliases from resources
+
+Given a corruption plan with Google calendar date of <date>
```
Comment (valfirst) [inline examples/core/src/main/java/org/jbehave/examples/core/stories/alias.story]:
```suggestion
Given a plan with Google calendar date of <date>
```

File: examples/core/src/main/resources/alias/aliases.json [hunk 3]
```diff
@@ -0,0 +1,21 @@
+[
+  {
+    "name": "Given a plan with calendar date of <date>",
+    "aliases" : [
+      {
+        "name": "Given a corruption plan with Google calendar date of <date>"
```
Comment (valfirst) [inline examples/core/src/main/resources/alias/aliases.json]:
```suggestion
        "name": "Given a plan with Google calendar date of <date>"
```

File: jbehave-core/src/main/java/org/jbehave/core/configuration/UnmodifiableConfiguration.java [hunk 4]
```diff
@@ -175,6 +181,11 @@ public Configuration useStoryParser(StoryParser storyParser) {
         throw notAllowed();
     }
 
+    @Override
+    public Configuration useAliasParser(AliasParser storyParser) {
+        throw notAllowed();
+    }
```
Comment (valfirst) [inline jbehave-core/src/main/java/org/jbehave/core/configuration/UnmodifiableConfiguration.java:193]: shouldn't `useAliasPaths` be overridden as well?

File: jbehave-core/src/main/java/org/jbehave/core/model/Variant.java [hunk 5]
```diff
@@ -0,0 +1,14 @@
+package org.jbehave.core.model;
+
+public class Variant {
```
Comment (valfirst) [inline jbehave-core/src/main/java/org/jbehave/core/model/Variant.java]:
```suggestion
public class AliasVariant {
```
?

File: jbehave-core/src/main/java/org/jbehave/core/parsers/JsonAliasParser.java [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+
+    private final Keywords keywords;
+
+    public JsonAliasParser(Keywords keywords) {
+        this.keywords = keywords;
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public Collection<Alias> parse(Set<String> resources) {
```
Comment (valfirst) [inline jbehave-core/src/main/java/org/jbehave/core/parsers/JsonAliasParser.java]:
```suggestion
    public Collection<Alias> parse(Set<String> aliasesAsStrings) {
```
?

File: jbehave-core/src/main/java/org/jbehave/core/parsers/JsonAliasParser.java [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+                        .collect(toList());
+    }
+
+    // https://github.com/google/gson/issues/61
+    private static UnaryOperator<AliasContainer> validate() {
+        return alias -> {
+            isTrue(alias.getName() != null, "The 'name' property that identifies step must be set");
+            Set<AliasEntry> aliases = alias.getAliases();
+            isTrue(aliases != null && !aliases.isEmpty(),
+                "The 'aliases' property is ether not set or point to an empty collection");
```
Comment (valfirst) [inline jbehave-core/src/main/java/org/jbehave/core/parsers/JsonAliasParser.java]:
```suggestion
                "The 'aliases' property is ether not set or points to an empty collection");
```

File: jbehave-core/src/main/java/org/jbehave/core/parsers/JsonAliasParser.java [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+                        .map(this::convert)
+                        .collect(toList());
+    }
+
+    // https://github.com/google/gson/issues/61
+    private static UnaryOperator<AliasContainer> validate() {
+        return alias -> {
+            isTrue(alias.getName() != null, "The 'name' property that identifies step must be set");
+            Set<AliasEntry> aliases = alias.getAliases();
+            isTrue(aliases != null && !aliases.isEmpty(),
```
Comment (valfirst) [inline jbehave-core/src/main/java/org/jbehave/core/parsers/JsonAliasParser.java]: I don't see any issue if empty collection of aliases is presented

**Review by valfirst — CHANGES_REQUESTED** [review]

**Review by valfirst — APPROVED** [review]

File: jbehave-core/src/main/java/org/jbehave/core/parsers/JsonAliasParser.java [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+
+        return aliasesAsStrings.stream()
+                               .map(json -> (List<AliasContainer>) gson.fromJson(json, aliasType))
+                               .flatMap(List::stream)
+                               .map(validate())
+                               .collect(groupingBy(AliasContainer::getName, collectingAndThen(toList(),
+                                       mergeAliases())))
+                               .values()
+                               .stream()
+                               .map(this::convert)
```
Comment (ikalinin1) [inline jbehave-core/src/main/java/org/jbehave/core/parsers/JsonAliasParser.java:50]: does it makes sense to follow same approach? either validate() convert() vs this::validate, this::convert?

**Review by ikalinin1 — APPROVED** [review]

**Comment by uarlouski** [comment]
In master https://github.com/jbehave/jbehave-core/commit/db96307be2dacb0abc0ba821809ab817b4c7a77d

**Pull request closed without merging**